### PR TITLE
zenitsu: DM curvature-weighted loss (per-point MSE weighting by local surface curvature)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

DrivAerML uses uniform MSE across all 50k surface points — every point contributes equally to the loss regardless of geometric complexity. But prediction errors concentrate at high-curvature regions: leading edges, mirror bases, wheel arches, A-pillar vortex lines. These regions have the steepest pressure gradients and are the hardest to predict.

By weighting the per-point MSE loss by local surface curvature (computed from the input normal vectors), we direct more gradient signal toward these challenging regions. This is different from:

- **Error-weighted SAMPLING** (jet #3243): changes which points are seen each epoch (requires error tracking from previous epoch)
- **Smoothness regularization** (emma #3231, dead): penalizes spatial gradients of predictions
- **This approach**: changes how much gradient each point contributes, using **GEOMETRY** (not prediction error) as the weighting signal

**Curvature estimation:** Surface curvature can be approximated from the normal field. For each point, compute the variance of normals among its K nearest neighbors. High normal variance = high curvature = more loss weight. This is a static geometric property — computed once per mesh, not updated during training.

---

## Baseline

Current best DrivAerML (PR #3072, eren/dm-ema-9995-gc05):
- `val_primary/surface_rel_l2_pct`: **3.833%** at epoch 511
- test: 4.685%
- W&B run: `ncl1dh88`
- Config: AdamW lr=5e-4, gc=0.5, no WD, 4L/512d/8H, EMA=0.9995, T_max=30, 394 batches/epoch

External target: **<3.71%** (AB-UPT). Gap: ~0.123 pp.

---

## Known Bug Fix

There is a `primary_metric_key` shadowing bug in `train.py` where a local variable (line ~1646-1648) shadows the function (line ~1428), causing an `UnboundLocalError`. Fix by renaming the local variable to `best_tracking_metric_key` if you encounter it.

---

## Implementation Guidance

1. Add `--curvature-loss-weight` flag (float, default `0.0` meaning uniform weighting)
2. For each mesh, **precompute** per-point curvature: for each of the 50k sampled points, find K=8 nearest neighbors and compute the standard deviation of their normal vectors. This gives a scalar curvature estimate per point.
3. Normalize curvature weights: `w = 1 + alpha * (curvature / curvature.mean())` where `alpha = curvature_loss_weight`. This ensures:
   - (a) flat regions still get weight 1.0
   - (b) high-curvature regions get extra weight proportional to their curvature
4. Apply per-point weighting to the MSE loss: `loss = (w * (pred - target)^2).mean()`
5. Log `mean_curvature_weight` and `max_curvature_weight` to W&B
6. Curvature can be precomputed during data loading — no per-step overhead beyond the weighted reduction

**IMPORTANT:** The KNN for curvature estimation is done ONCE during data loading, not every training step. This avoids the throughput penalty that killed the pressure gradient reg experiment.

---

### Trial 1 — `curvature_loss_weight=1.0` (moderate: high-curvature points get ~2x weight)

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --curvature-loss-weight 1.0 \
  --wandb_group dm-curvature-loss
```

### Trial 2 — `curvature_loss_weight=0.5` (lighter weighting)

Same as Trial 1 but with `--curvature-loss-weight 0.5`

---

## Deliverables

Report `val_primary/surface_rel_l2_pct` at best epoch for both trials.

**Target: beat 3.833%**

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT zenitsu:

## Results

### Summary

**Both trials failed to beat the baseline.** Curvature-weighted loss causes training instability on DrivAerML — both trials eventually diverged with exploding gradients. The hypothesis that upweighting high-curvature regions improves surface pressure prediction is not supported.

### Trial 1 — `curvature_loss_weight=1.0`

| Metric | Value |
|--------|-------|
| **val_primary/surface_rel_l2_pct** | **4.482%** (epoch 249) |
| test_primary/surface_rel_l2_pct | 5.448% |
| Total epochs | 461 (diverged at ~epoch 300) |
| mean_curvature_weight | 2.0 |
| max_curvature_weight | ~3.0 |
| W&B run | [`rwkptewj`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/rwkptewj) |

### Trial 2 — `curvature_loss_weight=0.5`

| Metric | Value |
|--------|-------|
| **val_primary/surface_rel_l2_pct** | **11.602%** (epoch 52) |
| test_primary/surface_rel_l2_pct | 12.143% |
| Total epochs | 460 (diverged at ~epoch 53) |
| mean_curvature_weight | 1.5 |
| max_curvature_weight | ~2.5 |
| W&B run | [`434enoo9`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/434enoo9) |

### Comparison vs Baseline

| | Baseline (uniform) | T1 (alpha=1.0) | T2 (alpha=0.5) |
|---|---|---|---|
| val_primary | **3.833%** (ep 511) | 4.482% (ep 249) | 11.602% (ep 52) |
| test_primary | **4.685%** | 5.448% | 12.143% |
| Δ val vs baseline | — | +0.649 pp | +7.769 pp |
| Diverged? | No | Yes (ep ~300) | Yes (ep ~53) |

### Training Commands

```bash
# Trial 1
CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --grad-clip 0.5 --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --ema-decay 0.9995 \
  --curvature-loss-weight 1.0 --agent zenitsu \
  --wandb-name "zenitsu/dm-curv-w1.0" --wandb-group dm-curvature-loss

# Trial 2 — identical except --curvature-loss-weight 0.5
CUDA_VISIBLE_DEVICES=1 python train.py \
  ... --curvature-loss-weight 0.5 ...
  --wandb-name "zenitsu/dm-curv-w0.5" --wandb-group dm-curvature-loss
```

### What Happened

**Curvature-weighted loss causes training instability and divergence on DrivAerML.**

Key observations:

1. **Both trials diverged with gradient explosion.** T2 (alpha=0.5) diverged much earlier (epoch 53) than T1 (alpha=1.0, epoch 300). This counter-intuitive ordering (lighter weighting diverges sooner) may be due to different random seeds in the cosine annealing interaction.

2. **Divergence pattern**: Loss suddenly jumps 4-10x, followed by gradient norms growing from ~1 to 2750+ (T2) or infinity (T1) within a few epochs. The model never recovers.

3. **Even before divergence, T1 plateaus at 4.48% — 17% worse than baseline (3.833%).** The curvature weighting fundamentally changes the loss landscape in a way that hurts the primary metric. The model over-fits to high-curvature regions at the expense of overall accuracy.

4. **Root cause analysis**: The curvature weights create a non-uniform gradient field where high-curvature points (leading edges, mirror bases) dominate the backward pass. Combined with cosine annealing's learning rate restarts (T_max=30), this creates periodic instability that eventually triggers divergence. The flat-region loss is suppressed relative to curvature peaks, so the model loses pressure field coherence in smooth regions that make up most of the surface.

5. **The KNN-based curvature computation** worked correctly (mean_weight=2.0 for alpha=1.0, 1.5 for alpha=0.5; max_weight~3.0). The implementation is sound — the hypothesis itself doesn't hold.

### Bug Fix

Also applied the `primary_metric_key` shadowing fix (consolidating dual best-tracking systems) from the EMA sweep PR.

### Suggested Follow-ups

1. **This direction should be abandoned.** Curvature-weighted loss is fundamentally at odds with the relative-L2 evaluation metric which treats all surface points equally.
2. If geometric weighting is revisited, consider **area-based weighting** (using the existing surface area column) as a free proxy — but given these results, any non-uniform weighting likely hurts.
3. **Error-weighted sampling** (jet #3243) may be more promising since it adapts based on prediction error, not static geometry.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3259 because ADVISOR: Closing — curvature-weighted loss is a clear dead end for DM. alpha=1.0: 4.482% best (+17%) then catastrophic divergence ep300. alpha=0.5: diverged ep53 (+203% best).

Root cause is two-fold: (1) **Train/eval mismatch** — curvature weighting optimizes a different objective than the uniform relative-L2 evaluation metric. High-curvature points dominate gradients while the metric weights all points equally. (2) **Gradient amplification at restarts** — curvature weights increase gradient variance, and cosine LR restarts (T_max=30) push the model off the weighted-loss basin into divergence. Alpha=0.5 diverged faster because the shallower basin was even less restart-tolerant.

Your observation about the structural misalignment between per-point weighting and the uniform metric is exactly right. Any static geometry-based loss weighting faces this same fundamental tension..
